Several fixes around POSIX_SOURCE, remove global definitions of POSIX_SOURCE

aescolar opened the pull request:
Some cleanup around the selection of the API provided by the C library.
The main idea is to limit where the SOURCE selection is applied to much fewer files than before.

This PR includes some commits from #67040 so as to provide a complete solution. But it only includes those commits that are inline with the overall direction of this PR of removing all global definitions of these macros.

Background on the PR strategy differences: 
Currently #67040 cleans up only the global definitions done by picolibc, but sets newlibc to define those globally.
The current assumption in 67040 is that all functions listed in [the coding guidelines rules A.4 and A.5](https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html#rule-a-5-c-standard-library-usage-restrictions-in-zephyr-codebase) should be globally available, without needing to set any macro by the user file, even that some of those functions are not part of the C std library API but extensions. Unfortunately, this sets a requirement on the libC integration to ensure they are exposed. See https://github.com/zephyrproject-rtos/zephyr/pull/67040#issuecomment-1896476000.

Instead, this PR solves the global use of POSIX_SOURCE issue, while postponing deciding on and solving the libC integration topic ( https://github.com/zephyrproject-rtos/zephyr/pull/67040#discussion_r1462390520 ) 

-------

Notes about the checkpatch warnings:
* For the native port specific driver "bottoms", coding guidelines rules A.4/5 do not really apply as they intend to limit the C library use in the embedded code, while this is host code. So I would think those warnings should just be ignored. A PR to clarify the rule has been submitted ( #68004 )
* The warnings in subsys/shell/modules/devmem_service.c & subsys/net/lib/lwm2m/lwm2m_rw_cbor.c are violations by the original code, which are uncovered now.
  * #68002
  * #68003
* The warning in tests/bluetooth/uuid/src/test_bt_uuid_to_str.c & tests/benchmarks/footprints/src/libc.c are due to strnlen which is an exception to rule A.4, and therefore I think the checkpatch warning in this PR should be waived.

Sidenote: We can avoid all those check-patch warnings by defining these macros in cmake, but it does not feel nice to do so, just because of checkpatch.

----

Fixes #66921

keith-packard committed:
libc/picolibc: Remove the global definition of _POSIX_C_SOURCE

This was necessary to get Picolibc to expose the whole Zephyr C library
API, but current versions of the SDK use a version of Picolibc with
built-in Zephyr support.

Signed-off-by: Keith Packard <keithp@keithp.com>

aescolar committed:
ztest native extensions: Set standard source macro appropriately

This file uses strtok_r which is an extension to the the
std C library. Let's explicity select one of the extensions
which includes it instead of relaying on somebody having
set it for this file somewhere else.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
tests rtc_api: Define _POSIX_C_SOURCE

This test uses functions which are extensions to
the C library. Let's explicity select one of the extensions
which includes it instead of relaying on somebody having
set it for this file somewhere else.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
drivers flash_simulator: Set standard source macro appropriately

This file uses several functions which are extensions to the the
std C library. Let's explicity select one of the extensions
which includes it instead of relaying on somebody having
set it for this file somewhere else.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
drivers epprom_simulator: Set standard source macro appropriately

This file uses several functions which are extensions to the the
std C library. Let's explicity select one of the extensions
which includes it instead of relaying on somebody having
set it for this file somewhere else.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
drivers uart_native_ptty: Set standard source macro appropriately

This file uses several functions which are extensions to the the
std C library. Let's explicity select one of the extensions
which includes it instead of relaying on somebody having
set it for this file somewhere else.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
board native_posix: Define required source standard macros

Instead of relaying on those macros having been defined
somewhere else let's define them for this library.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
arch posix: Do not define standard SOURCE macros globally

Do not define these macros globally, but instead
define them only for this library and when needed.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
subsys shell devmem_service: Fix includes

* The include for the host libC should be guarded with
  using the host libC instead of the POSIX arch, as this
  supports other C libraries.
* This code uses getopt, which is an extension to the
  C library. Let's explicity select one of the extensions
  which includes it instead of relaying on somebody having
  set it for this file somewhere else.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
drivers rtc: Do not treat the host libC differently

This code was using strptime from the C library in some cases,
but this function is an extension which many embedded libCs do
not provide, and which is not provided by default unless
_XOPEN_SOURCE or a similar macro is defined before the system headers
are included.

We could define _XOPEN_SOURCE for the libraries that provide it,
but instead of treating the host C library differently than
embedded libraries, let's just build the provided version always,
as that should provide better coverage of this code.

Note: It seems picolibc's strptime is broken,
until this very recent patch:
https://github.com/picolibc/picolibc/pull/657
so we should not select it when building for this library
for a while either.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
tests/lib/c_lib: Set standard source macro appropriately

strnlen() and strtok_r() are tested in this files
which are extensions to the the
std C library. Let's explicity select one of the extensions
also when building with the host C library,
instead of relaying on somebody having
set it for this file somewhere else.

Also, let's be nice and undefine them first in case they had
been defined somewhere else in the build scripts.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
tests modem/backends/tty: Define _XOPEN_SOURCE

This test uses functions which are extensions to
the C library. Let's explicity select one of the extensions
which includes it instead of relaying on somebody having
set it for this file somewhere else.
(Note this test is exclusive to native targets)

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
tests/benchmarks/footprints: Define required source standard macros

strnlen is not a C standard API, but an extension.
Instead of relaying that the SOURCE macro was set somewhere else
of that the C library provides a prototype anyhow
let's explicitly define it for this library.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
tests/bluetooth/uuid: Define required source standard macros

strnlen is not a C standard API, but an extension.
Instead of relaying that the SOURCE macro was set somewhere else
of that the C library provides a prototype anyhow
let's explicitly define it for this library.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
tests/net coap_client: Define _POSIX_C_SOURCE

This test uses functions and types which are extensions to
the C library. Let's explicity select one of the extensions
which includes it instead of relaying on somebody having
set it for this file somewhere else.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
subsys/net/lib/lwm2m: Define required source standard macros

Instead of relaying on those macros having been defined
somewhere else let's define them for this file.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

keith-packard committed:
posix: Use _POSIX_C_SOURCE=200809L instead of 200809

This constant is supposed to be defined as a long instead of an int,
presumably to support systems where int isn't large enough.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard commented on drivers/can/can_native_linux_adapt.c:
```diff
@@ -113,7 +113,7 @@ int linux_socketcan_poll_data(int fd)
 	return -EAGAIN;
 }
 
-ssize_t linux_socketcan_read_data(int fd, void *buf, size_t buf_len, bool *msg_confirm)
```
This change seems unrelated to the general topic, and also seems wrong -- the underlying issue is the incorrect cast of the ssize_t return from recvmsg to int.

aescolar replied:
They are related in the sense that ssize_t is an extension. So either ssize_t is not used, or this files set POSIX_C_SOURCE too (note the proto is a header used by "the embedded side").
It is correct that recvmsg (or write or read) can in theory return something in the whole range of ssize_t, but given the use, the maximum they will return is the size of the buffer sent or received which cannot be bigger than the packet size for the protocols which I do not know from the top of my head, but I'm guessing not bigger than IP's 64KB, so the int range should be more than sufficient.
That's why I just took the easier approach of simple having these return int (as the users anyhow also just cast to int), instead of adding POSIX_C_SOURCE to these embedded side of these drivers.

Does that sound sensible enough?

aescolar replied:
Another thing I forgot yesterday, these functions are built with one C library and called from another.
ssize_t being a C library defined type can in principle not match between both, so that would also speak for only using basic types defined by the compiler.
CC: @jukkar @henrikbrixandersen as authors of this code. Your input is also welcome :)

henrikbrixandersen replied:
I am fine with changing the function signature as this indeed could cause issues. To make the intention clear, perhaps add an explicit cast to int for the return value of the`recvmsg()` call?

keith-packard commented on drivers/eeprom/eeprom_simulator.c:
```diff
@@ -7,6 +7,20 @@
 #define DT_DRV_COMPAT zephyr_sim_eeprom
 
+#ifdef CONFIG_ARCH_POSIX
+#undef _POSIX_C_SOURCE
```
Why the undef here?

keith-packard reviewed (commented):
The ssize_t -> int changes are unrelated and also seem wrong? I'd also remove the `#undef` bits; they shouldn't be necessary once we get rid of the global definitions

keith-packard committed:
drivers/wifi/esp32: Remove _POSIX_C_SOURCE define as unneeded

This driver doesn't use any APIs outside of the Zephyr C library list, so
it doesn't need this _POSIX_C_SOURCE define.

Signed-off-by: Keith Packard <keithp@keithp.com>

bjarki-andreasen reviewed (commented):
I agree with the changes :)

cfriedt reviewed (changes requested):
Temporary nak until the issue described in #67181 can be brought to the AWG.

cfriedt reviewed (commented):
Thanks @aescolar - just a few suggestions / questions.

Currently I'm at 789/1242 tests complete and it looks reasonable.

I think this should unblock things though, at least in terms of global defines.

aescolar commented on boards/posix/native_posix/CMakeLists.txt:
```diff
@@ -3,6 +3,7 @@
 zephyr_library()
 
 zephyr_library_compile_definitions(NO_POSIX_CHEATS)
+zephyr_library_compile_definitions(_POSIX_C_SOURCE=200809 _XOPEN_SOURCE=600 _XOPEN_SOURCE_EXTENDED)
```
changed

aescolar commented on subsys/net/lib/lwm2m/lwm2m_rw_cbor.c:
```diff
@@ -7,6 +7,9 @@
 #define LOG_LEVEL CONFIG_LWM2M_LOG_LEVEL
 
+#undef _POSIX_C_SOURCE
+#define _POSIX_C_SOURCE 200809L /* Required for gmtime_r */
```
I would very much have that handled in a separate discussion as to allow this fix to progress.

aescolar commented on tests/lib/c_lib/strerror/src/main.c:
```diff
@@ -4,9 +4,8 @@
-#define _POSIX_C_SOURCE 200809
-#endif
+#undef _POSIX_C_SOURCE
+#define _POSIX_C_SOURCE 200809L /* Needed for strerror() which is tested here */
```
Given that the test, issue and workaround existed before this PR, I think it would be best to have this discussion happen outside of this PR and any possible fix be done in another context.

aescolar replied:
Same as https://github.com/zephyrproject-rtos/zephyr/pull/67052#discussion_r1466024433

aescolar replied:
The test was removed in https://github.com/zephyrproject-rtos/zephyr/pull/67891 . Therefore setting this macro is not necessary anymore.

aescolar commented on tests/bluetooth/uuid/src/test_bt_uuid_to_str.c:
```diff
@@ -1,6 +1,11 @@
 /* Copyright (c) 2022 Nordic Semiconductor ASA
  * SPDX-License-Identifier: Apache-2.0
  */
+#undef _POSIX_C_SOURCE
+/* Required for strnlen in some C libraries.
+ * Note strnlen is an exception to the coding guidelines rule A.4
+ */
+#define _POSIX_C_SOURCE 200809L
```
This proposal enters into the discussion from https://github.com/zephyrproject-rtos/zephyr/pull/67040#discussion_r1462390520 . Given that the current fix as is has no immediate drawbacks, and for the sake or allowing this PR to progress I would rather not re-start that discussion here again.

cfriedt replied:
Please make the requested change. Only newlib  and the host libc require these application conformance macros, and we really do not need the documentation repeated here.

aescolar replied:
In general it is not "only newlibc and the host libc" that need it, as much as "any C library which is not minimal or picolibc".
Why do you request this ifdefs be added?

aescolar replied:
@cfriedt Your original proposal has several changes:
a) You use CONFIG_HOST_LIBC but that does not exist I presume if anything that would  need to be EXTERNAL_LIBC
b) Please elaborate on why you want to `#if` these defines and if you will block this PR unless a similar `#if defined(..`  is done
c) Please clarify if you also will block this PR unless the comment is removed.

cfriedt replied:
> In general it is not "only newlibc and the host libc" that need it, as much as "any C library which is not minimal or picolibc".

Actually, there are certified ISO C libraries that do not provide the (POSIX) `strnlen()`.

>  Why do you request this ifdefs be added?

Because currently only the host libc and newlib require them. I believe this is self-explanatory. 

> The reason to mention explicitly which functions require this

I understand. However, it's covered in the documentation already. Having a single source of truth is nice (i.e. the documentation).

> is because we are tripping checkpatch

The solution in this PR does trip checkpatch.pl - there are alternatives that are probably more scalable and maintainable that do not trip checkpatch.pl, that do not involve hard-coding values in source files.. but that's a separate topic.

>  Please clarify if you also will block this PR unless the comment is removed.

I would prefer if the comment were removed.

aescolar replied:
>> Why do you request this ifdefs be added?

> Because currently only the host libc and newlib require them. I believe this is self-explanatory.

Unfortunately it is not self explanatory. You are requesting adding extra code to set conditionally a flag which can be set always. And I would prefer to not add unnecessary code.
So I would request you to reconsider rejecting this PR based on that.

aescolar commented on tests/lib/c_lib/common/src/main.c:
```diff
@@ -15,8 +15,13 @@
 
-#ifdef CONFIG_NEWLIB_LIBC
-#define _POSIX_C_SOURCE 200809
+#if defined(CONFIG_NEWLIB_LIBC) || defined(CONFIG_NATIVE_LIBC)
+/* strnlen() and strtok_r() are tested in this file. These are exceptions to coding
+ * guidelines rule A.5, but are not exposed by default by either Newlib or the host libC
+ * without setting this macro appropriately
+ */
+#undef _POSIX_C_SOURCE
+#define _POSIX_C_SOURCE 200809L
```
Same as https://github.com/zephyrproject-rtos/zephyr/pull/67052#discussion_r1466024433

cfriedt replied:
Please make the requested change. Only newlib  and the host libc require these application conformance macros, and we really do not need the documentation repeated here.

aescolar commented on subsys/shell/modules/devmem_service.c:
```diff
@@ -6,15 +6,24 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+/* This file uses getopt, which is only exposed if the appropriate macro is defined
```
> I think all of this can be reverted.

Each of this type of change was done to avoid a test failure

The answer from https://github.com/zephyrproject-rtos/zephyr/pull/67052#discussion_r1466024433 applies
It was agreed in the AWG to keep that RFC proposal and its discussion decoupled from this fix. Therefore I would request we do so.

cfriedt replied:
Yes, my original change request still applies - I guess remove the conditional I suggested if that makes you happier?

aescolar replied:
Is your request that the mention of the use of getopt and the note and link be removed?

cfriedt replied:
The comment is not necessary or necessarily accurate. `CONFIG_GETOPT` is used in several places within Zephyr.

Please remove it.

aescolar commented on tests/drivers/rtc/rtc_api/CMakeLists.txt:
```diff
@@ -34,3 +34,5 @@ if(DEFINED CONFIG_RTC_CALIBRATION)
 
 target_include_directories(app PRIVATE inc)
+
+target_compile_definitions(app PRIVATE _POSIX_C_SOURCE=200809L)
```
This is a test.
It was agreed in the AWG to keep that RFC proposal and its discussion decoupled from this fix. Therefore I would request we do so.

aescolar commented on tests/net/lib/coap_client/CMakeLists.txt:
```diff
@@ -16,6 +16,8 @@ target_sources(app PRIVATE ${ZEPHYR_BASE}/subsys/net/lib/coap/coap.c)
 target_include_directories(app PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/src)
 target_include_directories(app PRIVATE ${ZEPHYR_BASE}/include/)
 
+target_compile_definitions(app PRIVATE _POSIX_C_SOURCE=200809L)
```
Same answer as https://github.com/zephyrproject-rtos/zephyr/pull/67052#discussion_r1466032573
+
This test is specific for the native targets therefore, and setting the flag in general is harmless anyhow, so any filtering by target here would seem superfluous.

aescolar commented on tests/subsys/modem/backends/tty/CMakeLists.txt:
```diff
@@ -6,3 +6,4 @@ find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
 project(modem_backend_tty_test)
 
 target_sources(app PRIVATE src/main.c)
+target_compile_definitions(app PRIVATE _XOPEN_SOURCE=600)
```
This test and the module being tested are very specific for the native targets therefore, and setting the flag in general is harmless anyhow, so any filtering by target here would seem superfluous.

aescolar commented:
@cfriedt The comments that seemed to require fixing have been fixed. Please note that you are still rejecting/blocking this PR even that it is fixing a bug you reported.
Given that your remaining comments seem either suggestions or discussions about future changes, or enter into other currently ongoing discussions but which are not strictly necessary to do this fix, I would suggest the rejection to be removed.

cfriedt commented on tests/benchmarks/footprints/src/libc.c:
```diff
@@ -4,6 +4,12 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+#undef _POSIX_C_SOURCE
+/* Required for strnlen in some C libraries.
+ * Note strnlen is an exception to the coding guidelines rule A.4
+ */
+#define _POSIX_C_SOURCE 200809L
```
Please make the requested change. Only newlib  and the host libc require these application conformance macros, and we really do not need the documentation repeated here.

cfriedt commented:
> @cfriedt The comments that seemed to require fixing have been fixed. Please note that you are still rejecting/blocking this PR even that it is fixing a bug you reported.

Yes, I would still like to see some requested changes made.

Also, when you updated your code it was 3am for me. I sleep (sometimes).

> Given that your remaining comments seem either suggestions or discussions about future changes, or enter into other currently ongoing discussions but which are not strictly necessary to do this fix, I would suggest the rejection to be removed.

I've resolved a few comments. There are still some that I would like to see changed.

aescolar commented:
> There are still some that I would like to see changed.

@cfriedt Would you be so kind as to respond to the questions/clarification requests in your comments? That will probably speed up things.

cfriedt commented:
> > There are still some that I would like to see changed.
> 
> @cfriedt Would you be so kind as to respond to the questions/clarification requests in your comments? That will probably speed up things.

I've been doing that since 5:55am 😁 You do realize I'm not in your timezone, right?

henrikbrixandersen commented:
The compliance checks are false positives.

aescolar committed:
modules/zcbor: Fix implicit function declaration warning

Both zcbor_encode.c & zcbor_decode.c use strnlen()
In general these 2 functions are only exposed by the C library
if _POSIX_C_SOURCE is set 200809L.
But neither of these files (or their build scripts), are setting
this macro, causing build warnings
 Implicit declaration of function ‘strnlen’
which turn into failures in CI with some libCs.
Let's set this macro to avoid this issue.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
drivers/entropy/native: Define required SOURCE macro

The srandom function is not available
unless _XOPEN_SOURCE is set > 500 or an equivalent
declaration is done.
Let's set it.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

de-nordic reviewed (approved):
Looks OK. Thanks!

rlubos reviewed (approved):
Net part looks ok

cfriedt reviewed (approved):
LGTM - thanks @aescolar 👍
